core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -1219,6 +1219,44 @@ default void createMaterializedView(ConnectorSession session, SchemaTableName vi
         throw new TrinoException(NOT_SUPPORTED, "This connector does not support creating materialized views");
     }
 
+    /**
+     * Begin the creation of a materialized view with data.
+     */
+    default ConnectorInsertTableHandle beginCreateMaterializedView(
```

Comment:
Would this be called during analysis? Currently `BeginTableWrite` rule calls `beginInsert` for a table during planning. Ideally, `beginInsert` happens at beginning of query execution.

Couple of things here:
1. Currently `ConnectorMaterializedViewDefinition#storageTable` can point to a different connector, so `ConnectorInsertTableHandle` would break that feature
2. Even if storage table is from the same catalog. Then during `INSERT` analysis we perform `typesMatchForInsert` to support valid `INSERT` coercion. This might be useful if someone create MV that points to existing storage table or target 

The problem here is that we would like to (in sequence):
a. run analysis of `create MV`
b. perform `create MV`
c. run analysis of storage table
d. perform MV refresh

I don't think we have means for that. cc @martint 

Maybe if we give up on 1) and 2) we can make this work. Even then MV creation would have to be split into two steps:
1. prepare MV creation where neither MV or storage table is created, but MV (and maybe not-creted-yet storage table) are analyzed
2. actual execution step where both MV and storage tables are created.

It also makes me wonder if we could have things like `ConnectorTableHandle` or maybe (to be created) `ConnectorMaterializedViewHandle` be obtained in analysis for non-created objects yet.

Reply:
1. `beginCreateMaterializedView` will be called in `BeginTableWrite` rule similar to other operations such as CTAS.
2. The definition of `ConnectorMaterializedViewDefinition#storageTable` itself can point to a different catalog. However, in many other implementations, it is tightly coupled with the MV metadata catalog. To make it possible to have storage table in another connector, we may need separate SPIs for MV metadata creation & storage table creation. (Actually, having MV metadata & storage table in different connectors is needed for Athena, and it was already implemented. However, I found that it leads to an exception that multi-catalog writes in a transaction is not allowed. This limitation can be loosened for Athena, but I guess it is not possible for general Trino at least for now.)
3. The scenario that MV points to an existing storage table doesn't seem to make sense to me. The storage table is not controlled by a user but pure internal object. It looks more natural to generate it when a MV is created.
4. If you check CTAS implementation, the situation is similar. The analysis is performed for not-created object. It is not possible to do type checking for not-created object, but it just assumes that the table is properly created by a connector with given types.

Reply:
> > The definition of ConnectorMaterializedViewDefinition#storageTable itself can point to a different catalog.
> 
> I think this is something we should decide if we want to keep or not. For example, we have support for that in Starburst MVs. I don't think it's essential though. @claudiusli

Yes, definitely.
Anyhow, existing SPI is not sufficient to support storage table of different connector if it is newly created. (New SPI such as `createMaterializedViewStorageTable` or `beginCreateMaterializedViewStorageTable` would be needed.) The only possible scenario would be to point to existing storage table, but not really sure if it is valid scenario as mentioned in the previous comment. What would be the use case for that?
If pointing to an existing storage table is not a valid use case, would it be an option to support it later?

> > If you check CTAS implementation, the situation is similar. The analysis is performed for not-created object. It is not possible to do type checking for not-created object, but it just assumes that the table is properly created by a connector with given types.
> 
> This is useful for cases where target connector doesn't support all types of source connector. For example, in Starburst MVs we cast unsupported source types to vachar. Changing behavior will cause `CREATE MV` queries to fail (workaround would be `CREATE MV WITH NO DATA`)

Where is the type casting performed for Startburst MV? Is it performed inside refresh SPI? Then, I don't think this new interface does not change the existing behavior.